File: Roadmap/04 - CADENAS DE CARACTERES/php/qv1ko.php
<?php

    program("incapable", "display");

    function program($word1, $word2) {

        echo isPalindrome($word1) ? "The word $word1 is a palindrome\n" : "The word $word1 is not a palindrome\n";
        echo isPalindrome($word2) ? "The word $word2 is a palindrome\n" : "The word $word2 is not a palindrome\n";

        echo areAnagrams($word1, $word2) ? "The words $word1 and $word2 are anagrams\n" : "The words $word1 and $word2 are not anagrams\n";

        echo isIsogram($word1) ? "The word $word1 is an isogram\n" : "The word $word1 is not an isogram\n";
        echo isIsogram($word2) ? "The word $word2 is an isogram\n" : "The word $word2 is not an isogram\n";

    }

    function isPalindrome($word) {

        $len = strlen($word);

        for ($i = 0; $i < $len / 2; $i++) {
            if ($word[$i] != $word[$len - $i - 1]) {
                return false;
            }
        }

        return true;

    }

    function areAnagrams($word1, $word2) {

        if (strlen($word1) != strlen($word2)) {
            return false;
        }

        $lettersWord1 = str_split($word1);
        $lettersWord2 = str_split($word2);

        sort($lettersWord1);
        sort($lettersWord2);

        return $lettersWord1 == $lettersWord2;

    }

    function isIsogram($word) {

        $letters = [];

        foreach (str_split($word) as $letter) {
            if (in_array($letter, $letters)) {
                return false;
            }
            $letters[] = $letter;
        }

        return true;

    }

?>

File: Roadmap/04 - CADENAS DE CARACTERES/java/Qv1ko.java
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class Qv1ko {

    public static void main(String[] args) {
        program("provide", "seasonal");
    }

    private static void program(String word1, String word2) {

        System.out.println(isPalindrome(word1) ? "The word " + word1 + " is a palindrome" : "The word " + word1 + " is not a palindrome");
        System.out.println(isPalindrome(word2) ? "The word " + word2 + " is a palindrome" : "The word " + word2 + " is not a palindrome");

        System.out.println(areAnagrams(word1, word2) ? "The words " +  word1 + " and " +  word2 + " are anagrams" : "The words " +  word1 + " and " +  word2 + " are not anagrams");

        System.out.println(isIsogram(word1) ? "The word " + word1 + " is an isogram" : "The word " + word1 + " is not an isogram");
        System.out.println(isIsogram(word2) ? "The word " + word2 + " is an isogram" : "The word " + word2 + " is not an isogram");

    }

    private static boolean isPalindrome(String word) {

        int len = word.length();

        for (int i = 0; i < len / 2; i++) {
            if (word.charAt(i) != word.charAt(len - i - 1)) {
                return false;
            }
        }

        return true;

    }

    private static boolean areAnagrams(String word1, String word2) {

        if (word1.length() != word2.length()) {
            return false;
        }

        char[] lettersWord1 = word1.toCharArray();
        char[] lettersWord2 = word2.toCharArray();

        Arrays.sort(lettersWord1);
        Arrays.sort(lettersWord2);

        return Arrays.equals(lettersWord1, lettersWord2);

    }

    public static boolean isIsogram(String word) {

        Set<Character> letters = new HashSet<>();

        for (char letter : word.toCharArray()) {
            if (!letters.add(letter)) {
                return false;
            }
        }

        return true;

    }

}
